lib/compiler/src/beam_ssa_private_append.erl

```diff
@@ -0,0 +1,613 @@
+            Ps = keysort(1, map(ExtractOpargs, Patches)),
+            {Is, Cnt} = patch_opargs(I0, Ps, Cnt0),
+            patch_appends_is(Rest, PD, Cnt, Is++Acc, BlockAdditions0);
+        [{append,Dst,_}] ->
+            #b_set{op=bs_create_bin,dst=Dst,args=Args0}=I0,
+            [#b_literal{val=append}|OtherArgs] = Args0,
+            I = I0#b_set{args=[#b_literal{val=private_append}|OtherArgs]},
+            patch_appends_is(Rest, PD, Cnt0, [I|Acc], BlockAdditions0);
+        [{phi,Dst,_,_,_}|_] ->
+            {I, Extra, Cnt} = patch_phi(I0, Patches, Cnt0),
+            patch_appends_is(Rest, PD, Cnt, [I|Acc], Extra ++ BlockAdditions0)
+    end;
+patch_appends_is([I|Rest], PD, Cnt, Acc, BlockAdditions) ->
+    patch_appends_is(Rest, PD, Cnt, [I|Acc], BlockAdditions);
+patch_appends_is([], _, Cnt, Acc, BlockAdditions) ->
+    {reverse(Acc), Cnt, BlockAdditions}.
+
+%% The only time when we patch a return is when it returns a
+%% literal.
+patch_appends_ret(Last=#b_ret{arg=#b_literal{val=Lit}}, Patches, Cnt0)
+  when is_list(Lit); is_tuple(Lit) ->
+    Ps = keysort(1, [E || {ret,_,E} <- Patches]),
+    ?DP("patch_appends_ret tuple or list :~n  lit: ~p~n  patches: ~p~n", [Lit, Ps]),
+    {V,Extra,Cnt} = patch_literal_term(Lit, Ps, Cnt0),
+    {Last#b_ret{arg=V}, Extra, Cnt};
+patch_appends_ret(Last=#b_ret{arg=#b_literal{val=Lit}},
+                  [{ret,_,Element}],
+                  Cnt0) ->
+    ?DP("patch_appends_ret other:~n  lit: ~p~n  element: ~p~n", [Lit, Element]),
+    {V,Extra,Cnt} = patch_literal_term(Lit, Element, Cnt0),
+    {Last#b_ret{arg=V}, Extra, Cnt}.
+
+%% Should return the instructions in reversed order
+patch_opargs(I0=#b_set{args=Args}, Patches0, Cnt0) ->
+    ?DP("Patching args in ~p~nArgs: ~p~n Patches: ~p~n",
+        [I0,Args,Patches0]),
+    Patches = merge_arg_patches(Patches0),
+    {PatchedArgs, Is, Cnt} = patch_opargs(Args, Patches, 0, [], [], Cnt0),
+    {[I0#b_set{args=reverse(PatchedArgs)}|Is], Cnt}.
+
+patch_opargs([#b_literal{val=Lit}|Args], [{Idx,Lit,Element}|Patches],
+             Idx, PatchedArgs, Is, Cnt0) ->
+    ?DP("Patching arg idx ~p~n  lit: ~p~n  elem: ~p~n", [Idx, Lit, Element]),
+    {Arg,Extra,Cnt} = patch_literal_term(Lit, Element, Cnt0),
+    patch_opargs(Args, Patches, Idx + 1, [Arg|PatchedArgs], Extra ++ Is, Cnt);
+patch_opargs([Arg|Args], Patches, Idx, PatchedArgs, Is, Cnt) ->
+    ?DP("Skipping arg idx ~p~n  arg: ~p~n  patches: ~p~n",
+              [Idx,Arg,Patches]),
+    patch_opargs(Args, Patches, Idx + 1, [Arg|PatchedArgs], Is, Cnt);
+patch_opargs([], [], _, PatchedArgs, Is, Cnt) ->
+    {PatchedArgs, Is, Cnt}.
+
+%% The way find_defs and find_appends work, we can end up with
+%% multiple patches pacting different parts of a tuple or pair. We
+%% merge them here.
+merge_arg_patches([{Idx,Lit,P0},{Idx,Lit,P1}|Patches]) ->
+    P = case {P0, P1} of
+            {{hd,H},{tl,T}} ->
+                {pair,H,T};
+            {{tl,T},{hd,H}} ->
+                {pair,H,T};
```

Comment:
Not covered.

Reply:
Currently the type analyzer does not handle the case when you "misuse" cons cells. For example:
```Erlang
b(L) ->
    b(L, [<<>>|<<>>]).

b([H|T], [A|B]) ->
    b([H|T], [<<A/binary,H:8>>|<<B/binary,H:8>>]);
b([], Acc) ->
    Acc.
```
Here the type analyzer will conclude that the type of the accumulator is `nonempty_improper_list(bitstring(8,appendable), bitstring(8))` and not `nonempty_improper_list(bitstring(256,appendable), bitstring(256,appendable))` which would be expected. If the type analysis pass is fixed so both fields of the cons are transformed by private append, we will hit this code with the example above.

Apart from a couple of lines dealing with nifs (will fix that), all other non-covered lines in  [beam_ssa_private_append.erl](https://github.com/erlang/otp/pull/6804/files/0af6c67fc3dc4c5072bc77575e4c2cac7fea48b8#diff-4cdc13ddd828821dd63acadce3756539e7a875e7ee4a8f62d9d9f28fda35c5ae) are due to this type analysis fault as we never try to transform pairs of binaries.

When I last looked at this, during the autumn, I seem to recall that the type was less specific, i.e. `nonempty_improper_list(bitstring(8,appendable), any)` so maybe this can now be fixed. I'll have a look.

Reply:
This is now fixed in d1211f5b498615c999d147ed7a6c241469db19e0, `beam_ssa_type:type/5` had problems handling `bs_create_bin` when the
first fragment was typed as a #t_union{} containing an appendable bitstring. The same error was in beam_ssa_private_append:is_appendable/2` (fixed in 6952cb6310a06ea79e96f80ebdf0c1f68fda55a7).

Additionally 6952cb6310a06ea79e96f80ebdf0c1f68fda55a7 adds a tests which exercises `ht`/`tl` lines above, this getting the coverage to 99% (the remaining non-covered lines are due to the NIF problem mentioned below).

Reply:
` {{tl,T},{hd,H}}` never happens, patches are stored in an orddict, so only ` {{hd,H},{tl,T}}` can occur. So the non-reachable code is removed in b33d99e2dbb6a40a0941c02db6fc29e31dac3310.